File: apps/mobile/src/modules/photo-viewer/photoInfoFormatters.ts
import type { GalleryExif, GalleryFujiRecipe, GalleryToneAnalysis, GalleryToneType } from '@/modules/galleries/types'

export type Translator = (key: string, options?: Record<string, unknown>) => string

const dateFormatters = new Map<string, Intl.DateTimeFormat>()
const numberFormatters = new Map<string, Intl.NumberFormat>()
const signedFormatters = new Map<string, Intl.NumberFormat>()

export function textValue(value: unknown): string | null {
  if (typeof value === 'string') {
    return value.trim() || null
  }
  if (typeof value === 'number' && Number.isFinite(value)) {
    return String(value)
  }
  return null
}

export function translateExifValue(t: Translator, prefix: string, value: unknown): string | null {
  const text = textValue(value)
  if (!text) {
    return null
  }
  const suffix = text
    .toLowerCase()
    .replaceAll('&', 'and')
    .replaceAll(/[^a-z0-9]+/g, '-')
    .replaceAll(/^-|-$/g, '')
  return t(`${prefix}.${suffix}`, { defaultValue: text })
}

export function numberValue(value: unknown): number | null {
  if (typeof value === 'number') {
    return Number.isFinite(value) ? value : null
  }
  if (typeof value !== 'string' || !value.trim()) {
    return null
  }
  const parsed = Number(value)
  return Number.isFinite(parsed) ? parsed : null
}

export function formatNumber(value: number, locale: string, maximumFractionDigits = 1): string {
  const key = `${locale}:${maximumFractionDigits}`
  let formatter = numberFormatters.get(key)
  if (!formatter) {
    formatter = new Intl.NumberFormat(locale, { maximumFractionDigits })
    numberFormatters.set(key, formatter)
  }
  return formatter.format(value)
}

export function formatDate(value: string | null | undefined, locale: string): string | null {
  if (!value) {
    return null
  }
  const date = new Date(value)
  if (Number.isNaN(date.getTime())) {
    return value
  }
  let formatter = dateFormatters.get(locale)
  if (!formatter) {
    formatter = new Intl.DateTimeFormat(locale, { dateStyle: 'long', timeStyle: 'medium' })
    dateFormatters.set(locale, formatter)
  }
  return formatter.format(date)
}

export function formatFileSize(value: number | null, locale: string): string | null {
  if (value === null || !Number.isFinite(value) || value <= 0) {
    return null
  }

  const units = ['B', 'KB', 'MB', 'GB']
  let size = value
  let unitIndex = 0
  while (size >= 1024 && unitIndex < units.length - 1) {
    size /= 1024
    unitIndex += 1
  }
  return `${formatNumber(size, locale)} ${units[unitIndex]}`
}

export function formatMegapixels(width: number, height: number, locale: string): string | null {
  if (width <= 0 || height <= 0) {
    return null
  }
  return `${formatNumber((width * height) / 1_000_000, locale)} MP`
}

export function formatDimensions(width: number, height: number): string | null {
  return width > 0 && height > 0 ? `${width} × ${height}` : null
}

// exiftool hands back focal lengths and f-numbers as strings that may carry a
// trailing zero ("70.0"), which reads as noise next to Photos' own metrics.
function trimTrailingZero(value: number): string {
  return String(Number(value.toFixed(2)))
}

function focalLengthNumber(value: unknown): string | null {
  const numeric = numberValue(value)
  if (numeric !== null) {
    return trimTrailingZero(numeric)
  }
  const text = textValue(value)
  // exiftool already appends the unit ("70.0 mm"); re-parse so the trailing zero goes too.
  const withUnit = text?.match(/^(-?\d+(?:\.\d+)?)\s*mm$/i)
  return withUnit ? trimTrailingZero(Number(withUnit[1])) : null
}

export function formatFocalLength(value: unknown): string | null {
  const number = focalLengthNumber(value)
  if (number !== null) {
    return `${number} mm`
  }
  const text = textValue(value)
  if (!text) {
    return null
  }
  return /\bmm\b/i.test(text) ? text : `${text} mm`
}

/// Actual and 35mm-equivalent focal lengths are one fact in two unit systems, so
/// they share a cell — two bare `mm` values sitting apart read as two facts.
export function formatFocalPair(actual: unknown, equivalent: unknown): string | null {
  const actualNumber = focalLengthNumber(actual)
  const equivalentNumber = focalLengthNumber(equivalent)

  if (actualNumber !== null && equivalentNumber !== null && actualNumber !== equivalentNumber) {
    return `${actualNumber}→${equivalentNumber} mm`
  }
  return formatFocalLength(equivalent) ?? formatFocalLength(actual)
}

export function formatApertureGlyph(value: unknown): string | null {
  const numeric = numberValue(value)
  if (numeric !== null) {
    return `ƒ${trimTrailingZero(numeric)}`
  }
  const text = textValue(value)
  if (!text) {
    return null
  }
  return `ƒ${text.replace(/^f\//i, '')}`
}

export function formatExposureTime(value: unknown, locale: string): string | null {
  const numeric = numberValue(value)
  if (numeric !== null && numeric > 0) {
    if (numeric < 1) {
      return `1/${Math.round(1 / numeric)} s`
    }
    return `${formatNumber(numeric, locale, 2)} s`
  }

  const text = textValue(value)
  if (!text) {
    return null
  }
  return /\b(?:s|sec|seconds?)$/i.test(text) ? text : `${text} s`
}

export function formatEv(value: unknown): string | null {
  const text = textValue(value)
  if (!text) {
    return null
  }
  return /\bEV$/i.test(text) ? text : `${text} EV`
}

export function formatExposureBias(value: unknown, locale: string): string | null {
  const numeric = numberValue(value)
  if (numeric === null) {
    return null
  }
  let formatter = signedFormatters.get(locale)
  if (!formatter) {
    formatter = new Intl.NumberFormat(locale, { maximumFractionDigits: 1, signDisplay: 'exceptZero' })
    signedFormatters.set(locale, formatter)
  }
  return `${formatter.format(numeric)} ev`
}

export function formatIso(value: unknown): string | null {
  const text = textValue(value)
  return text ? `ISO ${text}` : null
}

export function formatMired(value: unknown): string | null {
  const text = textValue(value)
  if (!text) {
    return null
  }
  return /\bmired$/i.test(text) ? text : `${text} Mired`
}

export function joinMakeAndModel(makeValue: unknown, modelValue: unknown): string | null {
  const make = textValue(makeValue)
  const model = textValue(modelValue)
  if (make && model) {
    return model.toLowerCase().startsWith(make.toLowerCase()) ? model : `${make} ${model}`
  }
  return make || model
}

export function formatCoordinate(value: unknown, reference: unknown): string | null {
  const coordinate = textValue(value)
  if (!coordinate) {
    return null
  }
  const suffix = textValue(reference)
  const formatted = coordinate.includes('°') ? coordinate : `${coordinate}°`
  return suffix ? `${formatted} ${suffix}` : formatted
}

export function decimalCoordinate(value: unknown, reference: unknown, limit: number): number | null {
  const coordinate = numberValue(value)
  if (coordinate === null) {
    return null
  }

  const direction = textValue(reference)?.toUpperCase()
  const signedCoordinate
    = direction === 'S' || direction === 'SOUTH' || direction === 'W' || direction === 'WEST'
      ? -Math.abs(coordinate)
      : coordinate

  return Math.abs(signedCoordinate) <= limit ? signedCoordinate : null
}

export function formatAltitude(exif: GalleryExif | null): string | null {
  if (!exif) {
    return null
  }
  const altitude = textValue(exif.GPSAltitude)
  if (!altitude) {
    return null
  }
  const reference = textValue(exif.GPSAltitudeRef)
  const isBelowSeaLevel = exif.GPSAltitudeRef === 1 || reference?.toLowerCase().includes('below')
  const signedAltitude = isBelowSeaLevel && !altitude.startsWith('-') ? `-${altitude}` : altitude
  return /\bm$/i.test(signedAltitude) ? signedAltitude : `${signedAltitude} m`
}

export function cleanRecipeValue(value: unknown): string | null {
  const text = textValue(value)
  return text?.replace(/\s*\([^)]*\)$/, '').trim() || null
}

export function formatFilmMode(value: unknown): string | null {
  const filmMode = textValue(value)
  switch (filmMode) {
    case 'F0/Standard (Provia)': {
      return 'Provia'
    }
    case 'F1b/Studio Portrait Smooth Skin Tone (Astia)': {
      return 'Astia'
    }
    case 'F2/Fujichrome (Velvia)':
    case 'F4/Velvia': {
      return 'Velvia'
    }
    default: {
      return filmMode
    }
  }
}

export function formatFujiDynamicRange(recipe: GalleryFujiRecipe, t: Translator): string | null {
  if (recipe.DynamicRangeSetting === 'Manual' && recipe.DevelopmentDynamicRange) {
    return `DR${recipe.DevelopmentDynamicRange}`
  }
  if (recipe.DynamicRangeSetting === 'Auto') {
    return t('action.auto')
  }
  return textValue(recipe.DynamicRange)
}

export function formatFujiWhiteBalance(recipe: GalleryFujiRecipe, t: Translator): string | null {
  if (recipe.WhiteBalance === 'Kelvin' && recipe.ColorTemperature) {
    return `${recipe.ColorTemperature} K`
  }
  return translateExifValue(t, 'exif.fujirecipe-whitebalance', recipe.WhiteBalance)
}

export function formatPercentage(value: number, scale: number): string | null {
  return Number.isFinite(value) ? `${Math.round(value * scale)}%` : null
}

const toneTypeKeys: Record<GalleryToneType, string> = {
  'low-key': 'exif.tone.low-key',
  'high-key': 'exif.tone.high-key',
  'normal': 'exif.tone.normal',
  'high-contrast': 'exif.tone.high-contrast',
}

export function formatToneType(toneAnalysis: GalleryToneAnalysis | null, t: Translator): string | null {
  if (!toneAnalysis) {
    return null
  }
  return t(toneTypeKeys[toneAnalysis.toneType] ?? toneAnalysis.toneType)
}

File: apps/mobile/Afilmory/Core/Models/PhotoInfoFormatters.swift
import Foundation

enum PhotoInfoFormatters {
  static func text(_ value: JSONValue?) -> String? {
    guard let value else { return nil }
    switch value {
    case .string(let string):
      return string.trimmingToNil
    case .number(let number):
      guard number.isFinite else { return nil }
      return number.rounded(.towardZero) == number ? String(Int64(number)) : String(number)
    default:
      return nil
    }
  }

  static func text(_ value: String?) -> String? {
    value?.trimmingToNil
  }

  static func number(_ value: JSONValue?) -> Double? {
    guard let value else { return nil }
    switch value {
    case .number(let number):
      return number.isFinite ? number : nil
    case .string(let string):
      return Double(string.trimmingCharacters(in: .whitespacesAndNewlines))
    default:
      return nil
    }
  }

  static func translatedExifValue(
    prefix: String,
    value: JSONValue?
  ) -> String? {
    guard let text = text(value) else { return nil }
    let suffix = text.lowercased()
      .replacingOccurrences(of: "&", with: "and")
      .replacingOccurrences(of: "[^a-z0-9]+", with: "-", options: .regularExpression)
      .replacingOccurrences(of: "^-|-$", with: "", options: .regularExpression)
    return Bundle.main.localizedString(
      forKey: "\(prefix).\(suffix)",
      value: text,
      table: "ExifValues"
    )
  }

  static func formatNumber(_ value: Double, localeIdentifier: String, maximumFractionDigits: Int = 1) -> String {
    let formatter = NumberFormatter()
    formatter.locale = Locale(identifier: localeIdentifier)
    formatter.numberStyle = .decimal
    formatter.minimumFractionDigits = 0
    formatter.maximumFractionDigits = maximumFractionDigits
    return formatter.string(from: NSNumber(value: value)) ?? String(value)
  }

  static func formatDate(
    _ value: String?,
    localeIdentifier: String,
    timeZone: TimeZone
  ) -> String? {
    guard let value else { return nil }
    guard let date = PhotoDateParser.date(value, timeZone: timeZone) else { return value }
    let language = PhotoDateLanguage.resolve(localeIdentifier)
    let formatter = DateFormatter()
    formatter.locale = language.formattingLocale
    formatter.timeZone = timeZone
    switch language {
    case .english:
      formatter.dateFormat = "MMMM d, yyyy 'at' h:mm:ss\u{202F}a"
    case .simplifiedChinese, .hongKongChinese, .traditionalChinese:
      formatter.dateFormat = "yyyy年M月d日 HH:mm:ss"
    case .japanese:
      formatter.dateFormat = "yyyy年M月d日 H:mm:ss"
    case .korean:
      formatter.dateFormat = "yyyy년 M월 d일 a h:mm:ss"
    }
    return formatter.string(from: date)
  }

  static func formatFileSize(_ value: Double?, localeIdentifier: String) -> String? {
    guard var size = value, size.isFinite, size > 0 else { return nil }
    let units = ["B", "KB", "MB", "GB"]
    var index = 0
    while size >= 1024, index < units.count - 1 {
      size /= 1024
      index += 1
    }
    return "\(formatNumber(size, localeIdentifier: localeIdentifier)) \(units[index])"
  }

  static func formatMegapixels(_ width: Double, _ height: Double, localeIdentifier: String) -> String? {
    guard width > 0, height > 0 else { return nil }
    return "\(formatNumber(width * height / 1_000_000, localeIdentifier: localeIdentifier)) MP"
  }

  static func formatDimensions(_ width: Double, _ height: Double) -> String? {
    guard width > 0, height > 0 else { return nil }
    return "\(integerString(width)) × \(integerString(height))"
  }

  static func formatFocalLength(_ value: JSONValue?) -> String? {
    if let number = focalLengthNumber(value) {
      return "\(number) mm"
    }
    guard let text = text(value) else { return nil }
    return text.range(of: "\\bmm\\b", options: [.regularExpression, .caseInsensitive]) == nil
      ? "\(text) mm"
      : text
  }

  static func formatFocalPair(_ actual: JSONValue?, _ equivalent: JSONValue?) -> String? {
    let actualNumber = focalLengthNumber(actual)
    let equivalentNumber = focalLengthNumber(equivalent)
    if let actualNumber, let equivalentNumber, actualNumber != equivalentNumber {
      return "\(actualNumber)→\(equivalentNumber) mm"
    }
    return formatFocalLength(equivalent) ?? formatFocalLength(actual)
  }

  static func formatApertureGlyph(_ value: JSONValue?) -> String? {
    if let number = number(value) {
      return "ƒ\(trimTrailingZero(number))"
    }
    guard let text = text(value) else { return nil }
    return "ƒ\(text.replacingOccurrences(of: "^f/", with: "", options: [.regularExpression, .caseInsensitive]))"
  }

  static func formatExposureTime(_ value: JSONValue?, localeIdentifier: String) -> String? {
    if let number = number(value), number > 0 {
      if number < 1 {
        return "1/\(Int((1 / number).rounded())) s"
      }
      return "\(formatNumber(number, localeIdentifier: localeIdentifier, maximumFractionDigits: 2)) s"
    }
    guard let text = text(value) else { return nil }
    return text.range(of: "\\b(?:s|sec|seconds?)$", options: [.regularExpression, .caseInsensitive]) == nil
      ? "\(text) s"
      : text
  }

  static func formatEV(_ value: JSONValue?) -> String? {
    guard let text = text(value) else { return nil }
    return text.range(of: "\\bEV$", options: [.regularExpression, .caseInsensitive]) == nil ? "\(text) EV" : text
  }

  static func formatExposureBias(_ value: JSONValue?, localeIdentifier: String) -> String? {
    guard let number = number(value) else { return nil }
    let formatter = NumberFormatter()
    formatter.locale = Locale(identifier: localeIdentifier)
    formatter.numberStyle = .decimal
    formatter.maximumFractionDigits = 1
    let formatted = formatter.string(from: NSNumber(value: number)) ?? String(number)
    return "\(formatted) ev"
  }

  static func formatISO(_ value: JSONValue?) -> String? {
    text(value).map { "ISO \($0)" }
  }

  static func formatMired(_ value: JSONValue?) -> String? {
    guard let text = text(value) else { return nil }
    return text.range(of: "\\bmired$", options: [.regularExpression, .caseInsensitive]) == nil
      ? "\(text) Mired"
      : text
  }

  static func joinMakeAndModel(_ makeValue: JSONValue?, _ modelValue: JSONValue?) -> String? {
    let make = text(makeValue)
    let model = text(modelValue)
    if let make, let model {
      return model.lowercased().hasPrefix(make.lowercased()) ? model : "\(make) \(model)"
    }
    return make ?? model
  }

  static func formatCoordinate(_ value: JSONValue?, reference: JSONValue?) -> String? {
    guard let coordinate = text(value) else { return nil }
    let suffix = text(reference)
    let formatted = coordinate.contains("°") ? coordinate : "\(coordinate)°"
    return suffix.map { "\(formatted) \($0)" } ?? formatted
  }

  static func decimalCoordinate(_ value: JSONValue?, reference: JSONValue?, limit: Double) -> Double? {
    guard let coordinate = number(value) else { return nil }
    let direction = text(reference)?.uppercased()
    let signed = ["S", "SOUTH", "W", "WEST"].contains(direction) ? -abs(coordinate) : coordinate
    return abs(signed) <= limit ? signed : nil
  }

  static func formatAltitude(_ exif: GalleryExif?) -> String? {
    guard let exif, let altitude = text(exif["GPSAltitude"]) else { return nil }
    let reference = text(exif["GPSAltitudeRef"])
    let below = exif["GPSAltitudeRef"]?.number == 1 || reference?.lowercased().contains("below") == true
    let signed = below && !altitude.hasPrefix("-") ? "-\(altitude)" : altitude
    return signed.range(of: "\\bm$", options: [.regularExpression, .caseInsensitive]) == nil ? "\(signed) m" : signed
  }

  static func cleanRecipeValue(_ value: JSONValue?) -> String? {
    text(value)?.replacingOccurrences(of: "\\s*\\([^)]*\\)$", with: "", options: .regularExpression).trimmingToNil
  }

  static func formatFilmMode(_ value: JSONValue?) -> String? {
    switch text(value) {
    case "F0/Standard (Provia)": "Provia"
    case "F1b/Studio Portrait Smooth Skin Tone (Astia)": "Astia"
    case "F2/Fujichrome (Velvia)", "F4/Velvia": "Velvia"
    case let value?: value
    case nil: nil
    }
  }

  static func formatFujiDynamicRange(_ recipe: [String: JSONValue]) -> String? {
    if text(recipe["DynamicRangeSetting"]) == "Manual",
       let range = number(recipe["DevelopmentDynamicRange"]),
       range != 0
    {
      return "DR\(integerString(range))"
    }
    if text(recipe["DynamicRangeSetting"]) == "Auto" {
      return String(localized: "Auto")
    }
    return text(recipe["DynamicRange"])
  }

  static func formatFujiWhiteBalance(_ recipe: [String: JSONValue]) -> String? {
    if text(recipe["WhiteBalance"]) == "Kelvin", let temperature = text(recipe["ColorTemperature"]) {
      return "\(temperature) K"
    }
    return translatedExifValue(
      prefix: "exif.fujirecipe-whitebalance",
      value: recipe["WhiteBalance"]
    )
  }

  static func formatPercentage(_ value: Double, scale: Double) -> String? {
    value.isFinite ? "\(Int((value * scale).rounded()))%" : nil
  }

  static func formatToneType(_ tone: GalleryToneAnalysis?) -> String? {
    guard let tone else { return nil }
    switch tone.toneType {
    case "low-key": return String(localized: "Low Key")
    case "high-key": return String(localized: "High Key")
    case "normal": return String(localized: "Normal")
    case "high-contrast": return String(localized: "High Contrast")
    default: return tone.toneType
    }
  }

  private static func focalLengthNumber(_ value: JSONValue?) -> String? {
    if let number = number(value) {
      return trimTrailingZero(number)
    }
    guard let text = text(value),
          let match = text.firstMatch(pattern: "^(-?\\d+(?:\\.\\d+)?)\\s*mm$"),
          let number = Double(match)
    else { return nil }
    return trimTrailingZero(number)
  }

  private static func trimTrailingZero(_ value: Double) -> String {
    let rounded = (value * 100).rounded() / 100
    return rounded.rounded(.towardZero) == rounded ? integerString(rounded) : String(rounded)
  }

  private static func integerString(_ value: Double) -> String {
    String(Int64(value))
  }
}

private extension String {
  func firstMatch(pattern: String) -> String? {
    guard let expression = try? NSRegularExpression(pattern: pattern),
          let match = expression.firstMatch(in: self, range: NSRange(startIndex..., in: self)),
          match.numberOfRanges > 1,
          let range = Range(match.range(at: 1), in: self)
    else { return nil }
    return String(self[range])
  }
}
